ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,76 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/http-proxy/filters"
+	"github.com/hashicorp/golang-lru"
+)
+
+var log = golog.LoggerFor("ratelimiter")
+
+type ratelimiter struct {
+	hostPeriods          map[string]time.Duration
+	hostAccessesByClient *lru.Cache
+	mx                   sync.Mutex
+}
+
+// New creates a new rate limiting filter that only allows access to the hosts
+// listed in the given hostPeriods, and limits the periodicity of requests to
+// each host to the given duration. It limits the number of clients tracked to
+// the the numClients with the most recent activity.
+func New(numClients int, hostPeriods map[string]time.Duration) filters.Filter {
+	if numClients <= 0 {
+		numClients = 5000
+	}
+	// We can safely ignore the error, since the only thing that would cause an
+	// error is numClients <= 0
+	cache, _ := lru.New(numClients)
+	return &ratelimiter{
+		hostPeriods:          hostPeriods,
+		hostAccessesByClient: cache,
+	}
+}
+
+func (f *ratelimiter) Apply(resp http.ResponseWriter, req *http.Request, next filters.Next) error {
+	host, _, err := net.SplitHostPort(req.Host)
+	if err != nil {
+		host = req.Host
+	}
+	f.mx.Lock()
+	period := f.hostPeriods[host]
+	if period == 0 {
+		f.mx.Unlock()
```

Comment:
I mean to extract the operations which require the lock to a separate method, something like `func  (f *ratelimiter) check(host string, remoteAddr net.Addr) error`.